Mini2_test/View/Presenter/BadgesView.swift

```diff
@@ -110,19 +107,49 @@ struct BadgesView: View {
                         }
                     }
                     
-                    CustomButton(text: buttonText)
-                        .padding(.top, 20)
+                    Button{
+                        guard let selectedBadge = selectedBadge else
+                        { return }
+                        
+                        if selectedBadge.isUnlocked{
+                            assignBadgeToCat(selectedBadge)
+                        } else {
+                            buyBadge(badge: selectedBadge)
+                        }
+                    }label: {
+                        CustomButton(text: buttonText)
+                            .padding(.top, 20)
+                    }
+                    
                 }
                 Spacer()
             }
             .padding()
             .padding(.horizontal, 20)
+            .background(
+                Color.black.opacity(showAlert ? 0.3 : 0)
+                    .edgesIgnoringSafeArea(.all)
+            )
+            .overlay(
+                showAlert ? AnyView(customAlertView) : AnyView(EmptyView())
+            )
         }
     }
     
-    private var buttonText: String {
-        guard let selectedBadge = selectedBadge else { return "Select a Badge" }
-        return cat.badges.contains(where: { $0.id == selectedBadge.id }) ? "Wear This" : "Buy With \(selectedBadge.price) Coins"
+    private var customAlertView: some View {
+        CustomAlertView(title: "Not Enough Coins!", message: alertMessage, buttonText: "OK") {
+            showAlert = false
+        }
+    }
+    
+    private func buyBadge(badge: Badge){
+        if user[0].coin >= badge.price {
+            badge.isUnlocked = true
+            user[0].coin -= badge.price
+        } else {
+            alertMessage = "You need \(badge.price - user[0].coin) more coins to buy this badge."
+            showAlert = true
+        }
```

Comment:
separate logic to ViewModel

Reply:
keknya ini nanti aja, uda tny ka khoi masi gabisa. Nanti kalo sempet baru ubah lagi ke MVVM